### PR TITLE
#26 Add HarborRegistryClean

### LINKED ISSUE #26: Add Harbor Clean
Currently, `dockhand` has a `DTRClean`. Need to add a `HarborClean` to provide similar functionality

### PR DESCRIPTION
**Pull request by boxboatmatt**
Signed-off-by: Matthew DeVenny <matt@boxboat.com>

Add HarborRegistryClean to cleanup `dockhand` managed images stored in Harbor

Closes #26

### CODE REVIEW

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                    contentType: "APPLICATION_JSON"
+            )?.getContent()
+            def repo = Config.pipeline.readJSON(text: result.toString())
+            repo.each { imageRepo ->
+                repositories.add(imageRepo)
+            }
+        }
+        return repositories
+
+//        return Utils.resultOrTest(repositories,
```
Comment (boxboatmatt) [inline src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: Need to add this in for testing

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    String readRepositoryTags(Registry registry, String name) {
+        def requestURI = registry.getRegistryUrl() + registryAPIBase + "/repositories/${name}/tags"
+        def result = Config.pipeline.httpRequest(
+                url: requestURI,
+                authentication: registry.credential,
+                httpMode: 'GET',
+                contentType: "APPLICATION_JSON"
+        )?.getContent()
+        return result
+//        return Utils.resultOrTest(result, """
```
Comment (boxboatmatt) [inline src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: Need to add this in for testing

**Comment by boxboatmatt** [comment]
For some reason using `jsonSlurper` like in the `DtrClean` would cause
```
Caused: java.io.NotSerializableException: groovy.json.JsonSlurper
```
@jcantosz perhaps I should update `DtrClean` as well to use `readJson`

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+class HarborRegistryClean implements Serializable {
+
+    protected final String registryAPIBase = '/api'
+
+    boolean dryRun
+    int retentionDays
+    List<String> registryKeys
+
+    List<Object> readRepositories(Registry registry) {
+        def requestURL = registry.getRegistryUrl() + registryAPIBase + '/projects?page_size=100'
```
[the remark this answers is not in the document]
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: According to swagger 100 is the max - so yeah going to need to paginate

**Comment by jcantosz** [comment]
Is there enough shared logic that we could retain a single class and fill in particulars for paths from the config yaml -- like we do for git repos: https://github.com/boxboat/dockhand/blob/master/resources/com/boxboat/jenkins/config.example.yaml#L25-L28

I see some differences in `readRepositories` and `readRepositoryTags`. But haven't traced through it all yet.

**Comment by boxboatmatt** [comment]
> Is there enough shared logic that we could retain a single class and fill in particulars for paths from the config yaml -- like we do for git repos: https://github.com/boxboat/dockhand/blob/master/resources/com/boxboat/jenkins/config.example.yaml#L25-L28
> 
> I see some differences in `readRepositories` and `readRepositoryTags`. But haven't traced through it all yet.

They are pretty different in responses - I am also going to need to paginate all of these requests as Caleb pointed out

**Comment by jcantosz** [comment]
> For some reason using `jsonSlurper` like in the `DtrClean` would cause
> 
> ```
> Caused: java.io.NotSerializableException: groovy.json.JsonSlurper
> ```
> 
> @jcantosz perhaps I should update `DtrClean` as well to use `readJson`

Weird. Nothing jumps out at me as being very different. Yeah we can make those changes for DTRClean

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+//                            "metadata" : {
+//                            "public" : "true"
+//                        }
+//                        }
+//                ])
+    }
+
+    String readRepositoryTags(Registry registry, String name) {
+        def requestURI = registry.getRegistryUrl() + registryAPIBase + "/repositories/${name}/tags"
+        def result = Config.pipeline.httpRequest(
```
[the remark this answers is not in the document]
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: According to swagger it is not limited by page size

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                httpMode: 'GET',
+                contentType: "APPLICATION_JSON"
+        )?.getContent()
+
+        def projects = Config.pipeline.readJSON(text: result.toString())
+
+        def repositories = []
+
+        projects.each { project ->
+            requestURL = registry.getRegistryUrl() + registryAPIBase + "/repositories?project_id=${project.project_id}&page_size=100"
```
[the remark this answers is not in the document]
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: Updated to paginate

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (caleblloyd) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: You might want to double-check by running a query with the pagination params

File: src/com/boxboat/jenkins/library/docker/DTRClean.groovy [hunk 6]
```diff
@@ -51,88 +54,91 @@ class DTRClean implements Serializable {
...
-                    }
+                    ]
                 ]
-            }
-        """)
+            ])
     }
 
-    String readRepositoryTags(Registry registry, String namespace, String name) {
+    List<Object> readRepositoryTags(Registry registry, String namespace, String name) {
```
[the remark this answers is not in the document]
Comment (caleblloyd) [reply src/com/boxboat/jenkins/library/docker/DTRClean.groovy]: also I'd double-check if tags do in-fact use pagination even if it's not listed in Swagger... seems deeply rooted in Harbor's API code

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+    boolean dryRun
+    int retentionDays
+    List<String> registryKeys
+
+    private int projectPageSize = 100
+    private int repositoryPageSize = 100
+
+    List<Object> requestProjects(Registry registry, int page) {
+        def requestURL = registry.getRegistryUrl() + registryAPIBase + "/projects?page_size=${projectPageSize}&page=${page}"
```
[the remark this answers is not in the document]
Comment (caleblloyd) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: check my code - I wrote that one notepad didn't test it

File: src/com/boxboat/jenkins/library/docker/DTRClean.groovy (hunk 6 above)
[the remark this answers is not in the document]
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/DTRClean.groovy]: Verified on customer harbor that over 800 tags were returned in one request

File: src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+    }
+
+    List<Map<String, Object>>  requestRepositories(Registry registry, String projectId) {
+        return requestPaginated(registry, "/repositories", repositoryPageSize,  ["project_id": projectId])
+    }
+
+    List<Map<String, Object>>  readRepositories(Registry registry) {
+
+        def projects = requestProjects(registry)
+        Config.pipeline.echo "projects: ${projects}"
```
Comment (caleblloyd) [inline src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: Debug output need to be removed?

Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/HarborRegistryClean.groovy]: yup... I'll remove that

**Pull request merged**